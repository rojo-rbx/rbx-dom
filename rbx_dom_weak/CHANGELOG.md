# rbx_dom_weak Changelog

## Unreleased Changes
* Implemented `Default` for `WeakDom`, useful when using Serde or creating an empty `WeakDom`

## 2.6.0 (2023-10-03)
* Added `WeakDom::clone_multiple_into_external` that allows cloning multiple subtrees all at once into a given `WeakDom`, useful for preserving `Ref` properties that point across cloned subtrees ([#364])

[#364]: https://github.com/rojo-rbx/rbx-dom/pull/364

## 2.5.0 (2023-08-09)
* Fix potential stack overflow when creating or inserting into a `WeakDom`. ([#279])
* Added `InstanceBuilder::has_property` for checking if an `InstanceBuilder` defines a given property. ([#283])
* Added `WeakDom::clone_within` and `WeakDom::clone_into_external` for cloning instances into the same or a different `WeakDom`, respectively. ([#312])

[#279]: https://github.com/rojo-rbx/rbx-dom/pull/279
[#283]: https://github.com/rojo-rbx/rbx-dom/pull/283
[#312]: https://github.com/rojo-rbx/rbx-dom/pull/312

## 2.4.0 (2022-06-05)
* Added `WeakDom::into_raw` for enabling fast, non-tree-preserving transformations.
* Added `empty`, `with_class`, and `set_class` methods to `InstanceBuilder`.

## 2.3.0 (2021-10-11)
* Updated to rbx_types 1.3.

## 2.2.0 (2021-07-19)
* Updated to rbx_types 1.2.

## 2.1.0 (2021-07-02)
* Updated to rbx_types 1.1.

## 2.0.0 (2021-06-26) (yanked)
* Updated to rbx_types 1.0.

## 2.0.0-alpha.1 (2021-02-16)
This release is a major, breaking change that introduces many fixes and features.

* `RbxTree` was replaced with `WeakDom`
	* Changed `new` to accept an `InstanceBuilder`, which can contain children.
	* Renamed `get_root_id` to `root_ref`.
	* Renamed `get_instance` to `get_by_ref`.
	* Renamed `get_instance_mut` to `get_by_ref_mut`.
	* Renamed `move_instance` to `transfer`.
	* Renamed `set_parent` to `transfer_within`.
	* Renamed `insert_instance` to `insert`. This method now accepts an `InstanceBuilder`.
	* Removed `remove_instance`, replaced by `destroy` and `transfer`.
	* Removed `iter_all_ids`.
	* Removed `descendants`.
	* Added `root` and `root_mut` for accessing the root instance directly.
* `RbxInstanceProperties` was replaced with `InstanceBuilder`
	* This API is completely different, as it's now a builder. It is now much easier to construct instances.
* `RbxInstance` was replaced with `Instance`
	* This type now directly exposes its values instead of implementing `Deref` for another type.
* Moved types into the `rbx_types` crate, re-exported as `rbx_dom_weak::types`.
	* `RbxId` was replaced with `types::Ref`, which can now represent null referents.
* Added `DomViewer` API from rbx_dom_test to make testing instance trees easier.

## 1.10.1 (2019-12-18)
* Updated `base64`, `md5`, and `uuid` dependencies

## 1.10.0 (2019-09-15)
* Added `RbxTree::set_parent`, for moving instances within a single tree.

## 1.9.0 (2019-07-12)
* Changed `BrickColor` to be much more correct
	* All BrickColor values are now correctly available instead of just palette colors
	* Enum values now align with the `BrickColor.Number` property and `BrickColor.new` constructor
	* `BrickColor` no longer serializes to strings in human-readable Serde formats to avoid information loss. These colors have colliding names:
		* Rust
		* Lilac
		* Gold
		* Deep orange

## 1.8.2 (2019-07-10)
* Added conversion from `Int32` to `BrickColor`. This should fix serialization of `SpawnLocation` instances, which use `int` instead of `BrickColor` in at least the XML model format.

## 1.8.1 (2019-06-20)
* Fixed Serde being able to deserialize `RbxValue` and `RbxValueType`'s internal `__Nonexhaustive` value. This should also fix Serde recommending this value to users in error messages.
* Fixed `BrickColor` serialization to use Roblox color names

## 1.8.0 (2019-06-10)
* Added the `SharedString` type, which is used by Roblox to reduce redundant copies of binary buffers. ([#63](https://github.com/rojo-rbx/rbx-dom/pull/63))

## 1.7.0 (2019-05-14)
* Added conversions from `BrickColor` to `Color3` and `Color3uint8`. ([#46](https://github.com/rojo-rbx/rbx-dom/pull/46))
* Added conversions from `Color3` to `Color3uint8` and vice-versa.
* Changed the human readable serialization of `BinaryString` values to be base64-encoded. This makes JSON-encoded values much smaller. This changes the details of the unspecified serialization of rbx_dom_weak.
* `RbxValueConversion` now derives `PartialEq`.

## 1.6.0 (2019-05-12)
* Added `RbxTree::move_instance` API to move instances from one tree to another.
* Fixed `RbxTree::descendants` to no longer return the instance of the ID you give it. This may break code relying on this (broken) assumption, but was definitely a bug.
* `RbxValueConversion` now derives `Debug` and `Clone` ([#52](https://github.com/rojo-rbx/rbx-dom/issues/52))

## 1.5.0 (2019-05-06)
* Added support for `BrickColor` ([#29](https://github.com/rojo-rbx/rbx-dom/pull/29))
* Added `RbxValue::try_convert_ref`, which can be used to try to coerce values. ([#42](https://github.com/rojo-rbx/rbx-dom/pull/42))
* `Content` values can now be inferred from string literals.
* `RbxValueType` now derives Serde's `Serialize` and `Deserialize` traits

## 1.4.0 (2019-03-27)
* Added support for new types:
	* `NumberRange`
	* [#15](https://github.com/rojo-rbx/rbx-dom/pull/15):
		* `Int64`
		* `Float64`
	* [#20](https://github.com/rojo-rbx/rbx-dom/pull/20):
		* `Rect`
		* `Ray`
		* `ColorSequence`
		* `NumberSequence`

## 1.3.0 (2019-03-14)
* Added `sort_children_by_key` and `sort_children_unstable_by_key` methods to `RbxInstance` to reorder children safely

## 1.2.0 (2019-03-13)
* `RbxTree` and `RbxInstance` are now clonable without an `unimplemented` panic
	* Cloned trees preserve their IDs as-is and there's no public API yet to transplant instances between trees. This is mostly useful for comparing trees before/after a mutation, which we're using in Rojo.

## 1.1.0 (2019-03-11)
* Marked `PhysicalProperties` as `pub`
* Fixed `Serialize` impl for `UnresolvedRbxValue`
	* This removes `Serialize` from `AmbiguousRbxValue`, which should not break any real code.

## 1.0.0 (2019-03-01)
* `RbxValue` and `RbxValueType` can no longer be matched exhaustively, which enables adding new types without breaking code in the future
* Added support for `UDim`, `UDim2`, and `Content` value types
* `PhysicalProperties` is no longer a stub type

## 0.3.1 (2019-02-26)
* Added support for bare bool values when deserializing `UnresolvedRbxValue`
* Implemented `Serialize` for `UnresolvedRbxValue` and `AmbiguousRbxValue`
* Implemented `From<RbxValue>` for `UnresolvedRbxValue`

## 0.3.0 (2019-02-14)
* Renamed crate from `rbx_tree` to `rbx_dom_weak`
* Added support for `Ref` values ([#8](https://github.com/rojo-rbx/rbx-dom/pull/8))
* Added `UnresolvedRbxValue` and `AmbiguousRbxValue`, intended to be used alongside `rbx_reflection` to make specifying values less verbose.

## 0.2.0 (2019-01-25)
* Added new variants for `RbxValue`:
	* Int32
	* Float32
	* Enum
	* Vector2
	* Color3
	* Color3uint8
	* Vector3int16
	* Vector2int16
	* CFrame
	* PhysicalProperties (Stub)

## 0.1.0
* Initial release
* Supports `String`, `Bool`, and `Vector3` property values
