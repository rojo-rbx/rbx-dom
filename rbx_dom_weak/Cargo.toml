[package]
name = "rbx_dom_weak"
description = "Weakly-typed Roblox DOM implementation for Rust"
version = "2.4.0"
license = "MIT"
documentation = "https://docs.rs/rbx_dom_weak"
homepage = "https://github.com/rojo-rbx/rbx-dom"
repository = "https://github.com/rojo-rbx/rbx-dom.git"
readme = "README.md"
authors = ["Lucien Greathouse <me@lpghatguy.com>"]
edition = "2018"

[dependencies]
rbx_types = { version = "1.4.0", path = "../rbx_types", features = ["serde"] }

serde = "1.0.137"

[dev-dependencies]
insta = { version="1.14.1", features=["yaml"] }
