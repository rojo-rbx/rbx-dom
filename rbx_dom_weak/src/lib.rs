//! rbx_dom_weak is a common representation of the Roblox DOM for Rust. It's
//! designed to play nicely with the borrow checker and allows accessing
//! instances by ID in constant time.
//!
//! Constructing a new tree of instances is accomplished by first creating an
//! [`InstanceBuilder`] object that describes a tree of instances and then
//! wrapping it with an [`WeakDom`]:
//!
//! ```
//! use rbx_dom_weak::{InstanceBuilder, WeakDom};
//!
//! let dm = InstanceBuilder::new("DataModel");
//!
//! let mut dom = WeakDom::new(dm);
//!
//! println!("ID of DOM root is {:?}", dom.root_ref());
//! ```
//!
//! Once we have a tree, we can use [`WeakDom::insert`] and
//! [`WeakDom::get_by_ref`] to add instances to the tree and retrieve them.
//!
//! ```
//! use rbx_dom_weak::{InstanceBuilder, WeakDom};
//!
//! let mut dom = WeakDom::new(InstanceBuilder::new("DataModel"));
//!
//! // We can define properties using any type that can be converted to an
//! // rbx_dom_weak::types::Variant.
//! let http_service = InstanceBuilder::new("HttpService")
//!     .with_property("HttpEnabled", true);
//!
//! let http_service_id = dom.insert(dom.root_ref(), http_service);
//!
//! println!("HttpService has ID {:?}", http_service_id);
//! ```
//!
//! To change properties on an instance that's already present in the tree, use
//! [`WeakDom::get_by_ref_mut`]. Note that it isn't possible to add or remove
//! children through this method, use [`WeakDom::insert`] and
//! [`WeakDom::destroy`] instead.

#![deny(missing_docs)]

mod dom;
mod instance;
mod viewer;

pub use rbx_types as types;

pub use ahash::AHashMap;

pub use crate::{
    dom::WeakDom,
    instance::{Instance, InstanceBuilder},
    viewer::{DomViewer, ViewedInstance},
};

/// Helper trait that provides convenience methods for `AHashMap`.
pub trait HashMapExt {
    /// Constructs an empty map.
    fn new() -> Self;

    /// Constructs an empty map with at least the specified capacity.
    fn with_capacity(capacity: usize) -> Self;
}

impl<K, V> HashMapExt for AHashMap<K, V> {
    /// Creates an empty `AHashMap` using the default value for its hasher.
    fn new() -> Self {
        AHashMap::default()
    }

    /// Creates an empty `AHashMap` with at least the specified capacity using
    /// the default value for its hasher.
    fn with_capacity(capacity: usize) -> Self {
        AHashMap::with_capacity_and_hasher(capacity, Default::default())
    }
}
