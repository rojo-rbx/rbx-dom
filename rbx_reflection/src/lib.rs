mod class_tag;
mod database;
mod migration;
mod property_tag;
mod serde_util;

pub use class_tag::*;
pub use database::*;
pub use migration::*;
pub use property_tag::*;
