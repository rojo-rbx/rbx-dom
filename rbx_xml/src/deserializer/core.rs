use std::{collections::HashMap, io::BufRead};

use rbx_dom_weak::{
    types::{Ref, SharedString, Variant, VariantType},
    InstanceBuilder, WeakDom,
};
use rbx_reflection::{DataType, ReflectionDatabase};

use crate::{deserializer::conversions, property_descriptor::find_canonical_property_descriptor};

use super::{
    data_types,
    error::{DecodeError, ErrorKind},
    reader::{XmlData, XmlReader},
};

pub(crate) fn deserialize_file<R: BufRead>(
    mut reader: XmlReader<R>,
    config: DecodeConfig,
) -> Result<WeakDom, DecodeError> {
    log::trace!("beginning file deserialization");
    let mut roblox = reader.expect_start_with_name("roblox")?;
    let version = roblox.get_attribute("version")?;
    if version != "4" {
        // The error must return an owned string because we don't want to attach
        // a lifetime to errors
        return Err(ErrorKind::InvalidVersion(version).err());
    }

    let root = InstanceBuilder::new("DataModel");
    let root_ref = root.referent();
    let mut state = XmlState {
        dom: WeakDom::new(root),
        metadata: HashMap::new(),
        shared_strings: HashMap::new(),
        read_refs: HashMap::new(),
        ref_properties: Vec::new(),
        sstr_properties: Vec::new(),
    };

    loop {
        match reader.peek() {
            Some(Ok(event)) => match event {
                XmlData::ElementStart { name, .. } => match name.as_str() {
                    "Meta" => deserialize_metadata(&mut reader, &mut state)?,
                    "SharedStrings" => deserialize_sstr(&mut reader, &mut state)?,
                    "Item" => deserialize_item(&mut reader, &config, &mut state, root_ref)?,
                    // `MetaBreakpoints` and `DebuggerManager` also exist in
                    // some XML files but we don't support those in any
                    // capacity right now.
                    "External" => {
                        log::trace!("skipping External element");
                        reader.skip_element()?
                    }
                    _ => {
                        log::trace!("unexpected element {name}");
                        // Rust's won't allow use to reference 'name` after
                        // `reader` is borrowed, so we simply don't. We needed
                        // to clone the name anyway (to deref it), so it works
                        // out.
                        let name_clone = name.clone();
                        reader.expect_next()?;
                        return Err(ErrorKind::UnknownElement(name_clone).err());
                    }
                },
                XmlData::ElementEnd { name } if name == "roblox" => {
                    log::trace!("finishing deserialization");
                    break;
                }
                event => {
                    log::trace!("unexpected event {event:?}");
                    reader.expect_next()?;
                    return Err(ErrorKind::UnexpectedToken(reader.offset()).err());
                }
            },
            // This is safe to unwrap because `peek` guarantees we know
            // what `next` returns
            Some(Err(_)) => return Err(reader.next().unwrap().unwrap_err()),
            None => return Err(ErrorKind::UnexpectedEof.err()),
        }
    }
    log::debug!("Deserialized {} Instances", state.read_refs.len());

    log::debug!("Rewriting Referent properties");
    for (inst_referent, prop_name, prop_value) in state.ref_properties {
        log::trace!("rewriting {inst_referent}.{prop_name}");
        let inst = state.dom.get_by_ref_mut(inst_referent).unwrap();

        if inst.properties.get(&prop_name).is_none() {
            if let Some(value) = state.read_refs.get(&prop_value) {
                inst.properties.insert(prop_name, Variant::Ref(*value));
            }
        } else {
            return Err(ErrorKind::DuplicateProperty(prop_name).err());
        }
    }

    log::debug!("Rewriting SharedString properties");
    for (inst_referent, prop_name, prop_value) in state.sstr_properties {
        log::trace!("rewriting {inst_referent}.{prop_name}");
        let inst = state.dom.get_by_ref_mut(inst_referent).unwrap();

        if inst.properties.get(&prop_name).is_none() {
            if let Some(value) = state.shared_strings.get(&prop_value) {
                inst.properties
                    .insert(prop_name, Variant::SharedString(value.clone()));
            }
        } else {
            return Err(ErrorKind::DuplicateProperty(prop_name).err());
        }
    }

    // TODO: Validate property data

    Ok(state.dom)
}

fn deserialize_metadata<R: BufRead>(
    reader: &mut XmlReader<R>,
    state: &mut XmlState,
) -> Result<(), DecodeError> {
    log::trace!("decoding Meta");
    let mut element = reader.expect_start_with_name("Meta")?;
    let name = element.get_attribute("name")?;
    let value = reader.eat_text()?;
    reader.expect_end_with_name("Meta")?;

    log::debug!("Found metadata {name} = {value}");
    state.metadata.insert(name, value);
    Ok(())
}

fn deserialize_sstr<R: BufRead>(
    reader: &mut XmlReader<R>,
    state: &mut XmlState,
) -> Result<(), DecodeError> {
    log::trace!("decoding SharedStrings");

    reader.expect_start_with_name("SharedStrings")?;
    loop {
        match reader.peek() {
            Some(Ok(event)) => match event {
                XmlData::ElementStart { .. } => {
                    let mut sstr = reader.expect_start_with_name("SharedString")?;
                    let hash = sstr.get_attribute("md5")?;
                    let value = base64::decode(reader.eat_text()?)?;
                    reader.expect_end_with_name("SharedString")?;

                    log::debug!("Found SharedString {hash} ({} bytes)", value.len());
                    state.shared_strings.insert(hash, SharedString::new(value));
                }
                XmlData::ElementEnd { name } if name == "SharedStrings" => {
                    log::trace!("finishing SharedStrings decoding");
                    reader.next();
                    return Ok(());
                }
                event => {
                    log::trace!("unexpected event {event:?}");
                    reader.expect_next()?;
                    return Err(ErrorKind::UnexpectedToken(reader.offset()).err());
                }
            },
            Some(Err(_)) => return Err(reader.next().unwrap().unwrap_err()),
            None => return Err(ErrorKind::UnexpectedEof.err()),
        }
    }
}

fn deserialize_item<R: BufRead>(
    reader: &mut XmlReader<R>,
    config: &DecodeConfig,
    state: &mut XmlState,
    parent: Ref,
) -> Result<(), DecodeError> {
    let mut item = reader.expect_start_with_name("Item")?;
    let class = item.get_attribute("class")?;
    // !!Change in behavior!!
    // Previously, `referent` wasn't required, it now is
    let read_ref = item.get_attribute("referent")?;

    if config.strict_class_names {
        if let Some(database) = config.database {
            if !database.classes.contains_key(class.as_str()) {
                return Err(ErrorKind::UnknownClass(class, read_ref).err());
            }
        } else {
            return Err(ErrorKind::StrictWithoutDatabase("class names").err());
        }
    }
    log::debug!("Attempting to deserialize Instance {read_ref} of class {class}");

    let real_ref = state.dom.insert(parent, InstanceBuilder::new(class));
    let mut properties = HashMap::new();

    loop {
        match reader.peek() {
            Some(Ok(event)) => match event {
                XmlData::ElementStart { name, .. } => match name.as_str() {
                    "Properties" => {
                        deserialize_properties(reader, config, state, real_ref, &mut properties)?
                    }

                    "Item" => deserialize_item(reader, config, state, real_ref)?,
                    _ => {
                        log::trace!("unexpected element {name}");
                        let name_clone = name.clone();
                        reader.expect_next()?;
                        return Err(ErrorKind::UnknownElement(name_clone).err());
                    }
                },
                XmlData::ElementEnd { name } if name == "Item" => {
                    log::trace!("finishing Item decoding");
                    reader.next();
                    break;
                }
                event => {
                    log::trace!("unexpected event {event:?}");
                    reader.expect_next()?;
                    return Err(ErrorKind::UnexpectedToken(reader.offset()).err());
                }
            },
            Some(Err(_)) => return Err(reader.next().unwrap().unwrap_err()),
            None => return Err(ErrorKind::UnexpectedEof.err()),
        }
    }
    log::trace!("found {} properties", properties.len());
    let inst = state.dom.get_by_ref_mut(real_ref).unwrap();

    if let Some(value) = properties.remove("Name") {
        if let Variant::String(name) = value {
            inst.name = name
        } else {
            return Err(ErrorKind::NameNotString(value.ty()).err());
        }
    }

    // TODO make this... less strict?
    // This fails on impossible conversions, which is actually more restrictive
    // than what the old rbx_xml did. We'll have to either establish more
    // migrations or simply ignore any that can't happen.
    if config.strict_data_types || config.strict_property_names {
        if let Some(database) = config.database {
            for (prop_name, value) in properties.iter_mut() {
                let class_name = &inst.class;
                let canonical = find_canonical_property_descriptor(database, class_name, prop_name);
                match canonical {
                    Some(descriptor) => {
                        let canonical_type = match &descriptor.data_type {
                            DataType::Value(ty) => *ty,
                            DataType::Enum(_) => VariantType::Enum,
                            _ => unimplemented!(),
                        };
                        if canonical_type != value.ty() {
                            log::debug!("Attempting to convert {class_name}.{prop_name} to be of type {canonical_type:?} (currently {:?})", value.ty());
                            if let Err(error) = conversions::convert(value, canonical_type) {
                                log::error!("Could not convert {value:?} to {canonical_type:?} because {error}");
                                // return Err(ErrorKind::ConversionFail {
                                //     class: class_name.clone(),
                                //     name: prop_name.clone(),
                                //     from: value.ty(),
                                //     to: canonical_type,
                                //     error,
                                // }
                                // .err());
                            }
                        }
                    }
                    None => {
                        if config.strict_property_names {
                            return Err(ErrorKind::UnknownProperty(
                                class_name.to_owned(),
                                prop_name.to_owned(),
                            )
                            .err());
                        }
                    }
                }
            }
        } else {
            return Err(ErrorKind::StrictWithoutDatabase("properties").err());
        }
    }

    inst.properties = properties;

    state.read_refs.insert(read_ref, real_ref);

    Ok(())
}

fn deserialize_properties<R: BufRead>(
    reader: &mut XmlReader<R>,
    config: &DecodeConfig,
    state: &mut XmlState,
    referent: Ref,
    properties: &mut HashMap<String, Variant>,
) -> Result<(), DecodeError> {
    log::trace!("decoding Properties");
    reader.expect_start_with_name("Properties")?;

    loop {
        match reader.peek() {
            Some(Ok(event)) => match event {
                XmlData::ElementStart { name, .. } => {
                    // We may be able to get around this later but for now
                    // we have to take ownership of the element name.
                    let prop_type = name.clone();
                    if data_types::is_known_type(&prop_type) {
                        let mut element = reader.expect_start_with_name(&prop_type)?;
                        let prop_name = element.get_attribute("name")?;
                        if log::log_enabled!(log::Level::Debug) {
                            let class_name = &state.dom.get_by_ref(referent).unwrap().class;
                            log::debug!("Attempting to deserialize property {prop_name} of type {prop_type}");
                        }
                        let variant = data_types::attempt_deserialization(reader, &prop_type)?;

                        if prop_type == "Ref" {
                            log::trace!("referent property {prop_name} = {variant:?}");
                            state.ref_properties.push((
                                referent,
                                prop_name,
                                match variant {
                                    Variant::String(str) => str,
                                    _ => unreachable!(),
                                },
                            ));
                        } else if prop_type == "SharedString" {
                            log::trace!("SharedString property {prop_name} = {variant:?}");
                            state.sstr_properties.push((
                                referent,
                                prop_name,
                                match variant {
                                    Variant::String(str) => str,
                                    _ => unreachable!(),
                                },
                            ));
                        } else if properties.get(&prop_name).is_none() {
                            properties.insert(prop_name, variant);
                        } else {
                            return Err(ErrorKind::DuplicateProperty(prop_name).err());
                        }

                        reader.expect_end_with_name(&prop_type)?;
                    } else {
                        log::warn!("Unknown property type {prop_type}");
                        if config.ignore_new_types {
                            reader.skip_element()?;
                        } else {
                            return Err(ErrorKind::UnknownType(prop_type).err());
                        }
                    }
                }
                XmlData::ElementEnd { name } if name == "Properties" => {
                    log::trace!("finished decoding properties");
                    reader.next();
                    return Ok(());
                }
                event => {
                    log::trace!("unexpected event {event:?}");
                    reader.expect_next()?;
                    return Err(ErrorKind::UnexpectedToken(reader.offset()).err());
                }
            },
            Some(Err(_)) => return Err(reader.next().unwrap().unwrap_err()),
            None => return Err(ErrorKind::UnexpectedEof.err()),
        }
    }
}

#[derive(Debug)]
struct XmlState {
    /// The internal WeakDom used by the decoder
    dom: WeakDom,
    /// A map of metadata values deserialized from `Meta` elements
    metadata: HashMap<String, String>,
    /// A map of SharedString values deserialized from the file
    shared_strings: HashMap<String, SharedString>,
    /// A map of all read Instance referents to their real referents
    read_refs: HashMap<String, Ref>,
    /// A list of properties that point to other Instances in the file.
    /// They need to be revisited after we're done deserializing so they can be
    /// rewritten to use our actual referents.
    ///
    /// The tuple is `(InstanceBuilder Referent, Property Name, Read Value)`
    ref_properties: Vec<(Ref, String, String)>,
    /// A list of properties that point to a SharedString. They need to be
    /// revisited after we're done deserializing so we can actually point to
    /// them.
    ///
    /// The tuple is `(Instance Referent, Property Name, Read Value)`
    sstr_properties: Vec<(Ref, String, String)>,
}

/// A struct configuring the behavior of the deserializer.
/// By default, this uses no database. To add one, use `set_database`.
#[derive(Debug)]
pub struct DecodeConfig<'db> {
    /// What database if any to use for decoding properties and classes.
    pub(crate) database: Option<&'db ReflectionDatabase<'db>>,
    /// When `true`, class names be checked against the database and
    /// an error will be raised when an unknown class is encountered.
    pub(crate) strict_class_names: bool,
    /// When `true`, property types will be checked against the database and
    /// an error will be raised when a type mismatch is encountered.
    pub(crate) strict_data_types: bool,
    /// When `true`, property names will be checked against the database and
    /// an error will be raised when unknown properties are encountered.
    pub(crate) strict_property_names: bool,
    /// When `true`, any new property data types will be skipped.
    /// Otherwise, an error will be raised when a new data type is encountered.
    pub(crate) ignore_new_types: bool,
}

impl<'db> Default for DecodeConfig<'db> {
    fn default() -> Self {
        Self {
            database: None,
            strict_class_names: false,
            strict_data_types: false,
            strict_property_names: false,
            // This is why we manually implement `Default`!
            ignore_new_types: true,
        }
    }
}

impl<'db> DecodeConfig<'db> {
    /// Creates a new `DecodeConfig` with the default options. This means
    /// no database is used and unknown data types are skipped during
    /// deserialization.
    pub fn new() -> Self {
        Self::default()
    }

    /// Creates a new `DecodeConfig` with the given database. By default,
    /// class names, property names, and property types are checked.
    /// Additionally, new data types are ignored.
    pub fn with_database(database: &'db ReflectionDatabase<'db>) -> Self {
        Self {
            database: Some(database),
            strict_class_names: true,
            strict_data_types: true,
            strict_property_names: false,
            ignore_new_types: true,
        }
    }

    /// Sets the deserializer to use the given `ReflectionDatabase`.
    pub fn database(mut self, database: &'db ReflectionDatabase<'db>) -> Self {
        self.database = Some(database);
        self
    }

    /// Sets whether class names are checked against the database. If `true`,
    /// an error will be raised during deserialization if an unknown class
    /// is encountered.
    pub fn strict_class_names(mut self, ignore: bool) -> Self {
        self.strict_class_names = ignore;
        self
    }

    /// Sets whether property data types are checked against the database.
    /// If `true`, an error will be raised during deserialization if a
    /// property's type does not match in the database.
    pub fn strict_data_types(mut self, ignore: bool) -> Self {
        self.strict_data_types = ignore;
        self
    }

    /// Sets whether property names are checked against the database.
    /// If `true`, an error will be raised during deserialization if a
    /// property is not known by name to the database.
    pub fn strict_property_names(mut self, ignore: bool) -> Self {
        self.strict_property_names = ignore;
        self
    }

    /// Sets whether unknown property data types are ignored during
    /// deserialization. If `true`, any property of an unknown type will be
    /// skipped.
    pub fn ignore_new_types(mut self, ignore: bool) -> Self {
        self.ignore_new_types = ignore;
        self
    }
}

#[cfg(test)]
mod tests {
    use rbx_dom_weak::DomViewer;

    use super::*;

    #[test]
    fn decode_test() {
        #![allow(unused_must_use)]
        env_logger::try_init();
        let document = r#"
            <roblox version="4">
                <External>TestExternal</External>
                <Meta name="TestMetadata">TestValue</Meta>
                <Item class="TestClass" referent="TestReferent">
                    <Properties>
                        <string name = "Name">"Test Name"</string>
                        <Ref name = "RefTest">null</Ref>
                        <SharedString name="TestSharedString">Test Shared String Key</SharedString>
                        <bool name="TestBool1">true</bool>
                        <bool name="TestBool2">false</bool>
                        <string name="TestString">Test Value</string>
                        <float name="TestFloat1">-0.15625</float>
                        <float name="TestFloat2">INF</float>
                        <float name="TestFloat3">-INF</float>
                        <float name="TestFloat4">NAN</float>
                        <double name="TestDouble1">-0.15625</double>
                        <double name="TestDouble2">INF</double>
                        <double name="TestDouble3">-INF</double>
                        <double name="TestDouble4">NAN</double>
                        <Vector3 name="TestVector3">
                            <X>1337</X>
                            <Y>123456789.10</Y>
                            <Z>-4276993775</Z>
                        </Vector3>
                    </Properties>
                    <Item class="TestClass2" referent="TestReferent2">
                        <Properties>
                            <Ref name = "RefTest">TestReferent</Ref>
                            <SharedString name="TestSharedString">Test Shared String Key</SharedString>
                            <int name = "TestInt1">10</int>
                            <int name = "TestInt2">-10</int>
                            <int64 name = "TestInt64_1">20</int64>
                            <int64 name = "TestInt64_2">-20</int64>
                            <ProtectedString name = "Test"><![CDATA[   Protected String   ]]></ProtectedString>
                            <Ray name="TestRay">
                                <origin>
                                    <X>1</X>
                                    <Y>2</Y>
                                    <Z>3</Z>
                                </origin>
                                <direction>
                                    <X>-4</X>
                                    <Y>-5</Y>
                                    <Z>-6</Z>
                                </direction>
                            </Ray>
                        </Properties>
                    </Item>
                </Item>
                <SharedStrings>
                    <SharedString md5="Test Shared String Key">Q1NHSzg1MTYxZjdlOWNmZjMyNTlhNmU1NmE2NGJjZmNjMzJh</SharedString>
                </SharedStrings>
            </roblox>
        "#;

        match deserialize_file(XmlReader::from_str(document), Default::default()) {
            Err(err) => panic!("{}", err),
            Ok(dom) => {
                insta::assert_yaml_snapshot!(
                    "deserializer feature list",
                    DomViewer::new().view(&dom)
                )
            }
        }
    }

    #[test]
    fn crossroads_decode() {
        #![allow(unused_must_use)]
        env_logger::try_init();
        let file = std::fs::File::open("benches/crossroads.rbxlx").unwrap();

        let reader = XmlReader::from_reader(std::io::BufReader::new(file));
        if let Err(err) = deserialize_file(reader, Default::default()) {
            panic!("{}", err)
        }
    }

    #[test]
    fn crossroads_strict() {
        #![allow(unused_must_use)]
        env_logger::try_init();
        let file = std::fs::File::open("benches/crossroads.rbxlx").unwrap();

        let reader = XmlReader::from_reader(std::io::BufReader::new(file));
        if let Err(err) = deserialize_file(
            reader,
            DecodeConfig::with_database(rbx_reflection_database::get()),
        ) {
            panic!("{}", err)
        }
    }

    #[test]
    fn decode_strict() {
        #![allow(unused_must_use)]
        env_logger::try_init();
        let document = r#"
            <roblox version="4">
                <Item class="Workspace" referent="TestReferent">
                    <Properties>
                        <string name = "Name">Test Workspace</string>
                    </Properties>
                    <Item class="Part" referent="TestReferent2">
                        <Properties>
                            <string name = "Name">Test Part</string>
                            <int name = "BrickColor">137</int>
                        </Properties>
                    </Item>
                </Item>
            </roblox>
        "#;

        match deserialize_file(
            XmlReader::from_str(document),
            DecodeConfig::with_database(rbx_reflection_database::get()),
        ) {
            Err(err) => panic!("{}", err),
            Ok(dom) => {
                insta::assert_yaml_snapshot!(
                    "deserialize with database",
                    DomViewer::new().view(&dom)
                )
            }
        }
    }
}
