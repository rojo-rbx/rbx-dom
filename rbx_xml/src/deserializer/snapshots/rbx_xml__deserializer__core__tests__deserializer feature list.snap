---
source: rbx_xml/src/deserializer/core.rs
expression: "DomViewer::new().view(&dom)"
---
referent: referent-0
name: DataModel
class: DataModel
properties: {}
children:
  - referent: referent-1
    name: "\"Test Name\""
    class: TestClass
    properties:
      RefTest: "null"
      TestAxes1:
        Axes: []
      TestAxes2:
        Axes:
          - Z
      TestAxes3:
        Axes:
          - X
          - Y
          - Z
      TestBool1:
        Bool: true
      TestBool2:
        Bool: false
      TestDouble1:
        Float64: -0.15625
      TestDouble2:
        Float64: inf
      TestDouble3:
        Float64: -inf
      TestDouble4:
        Float64: NaN
      TestFloat1:
        Float32: -0.15625
      TestFloat2:
        Float32: inf
      TestFloat3:
        Float32: -inf
      TestFloat4:
        Float32: NaN
      TestSharedString:
        len: 36
        hash: a8e881ce93449542139935bd0ec3f0a77725c368b5a699b2b87077d9fd842a64
      TestString:
        String: Test Value
      TestVector3:
        Vector3:
          - 1337
          - 123456790
          - -4276993800
    children:
      - referent: referent-2
        name: TestClass2
        class: TestClass2
        properties:
          RefTest: referent-1
          TestCFrame:
            CFrame:
              position:
                - 1
                - 2
                - 3
              orientation:
                - - inf
                  - NaN
                  - -inf
                - - -1
                  - -2
                  - -3
                - - 0.15625
                  - -0.5
                  - 0
          TestColor3:
            Color3:
              - 1
              - 0.7058824
              - 0.078431375
          TestColor3uint8:
            Color3uint8:
              - 163
              - 162
              - 165
          TestFaces1:
            Faces: []
          TestFaces2:
            Faces:
              - Right
              - Top
              - Back
              - Left
              - Bottom
          TestFaces3:
            Faces:
              - Right
              - Top
              - Back
              - Bottom
              - Front
          TestInt1:
            Int32: 10
          TestInt2:
            Int32: -10
          TestInt64_1:
            Int64: 20
          TestInt64_2:
            Int64: -20
          TestProtectedString:
            String: "   Protected String   "
          TestRay:
            Ray:
              origin:
                - 1
                - 2
                - 3
              direction:
                - -4
                - -5
                - -6
          TestSharedString:
            len: 36
            hash: a8e881ce93449542139935bd0ec3f0a77725c368b5a699b2b87077d9fd842a64
        children: []

